### PR TITLE
[REF] separated calculating of discounts for credits via discounts pa…

### PR DESCRIPTION
**Pull request by KolushovAlexandr**
…yments

### CODE REVIEW

**Comment by yelizariev** [comment]
Is it ready to review?

**Comment by KolushovAlexandr** [comment]
@yelizariev No, I know the scenario that doesn't work on

File: pos_debt_notebook/models.py [hunk 1]
```diff
@@ -465,11 +465,14 @@ def _compute_product_list(self):
     @api.model
     def _process_order(self, pos_order):
         credit_updates = []
+        amount_via_discount = 0
+        print pos_order
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: remove please

File: pos_debt_notebook/static/src/js/pos.js [hunk 2]
```diff
@@ -431,23 +430,23 @@ odoo.define('pos_debt_notebook.pos', function (require) {
...
+//                        if (discount_credits > 0.00001) {
+//                            old_price = ol.get_price_with_tax();
+//                            percentage = Math.min(Math.max(( discount_credits / old_price ) * 100, 0), 100);
+//                            ol.set_discount(percentage);
+//                            discount_credits -= old_price - ol.get_price_with_tax();
+//                        } else {
+//                            return;
+//                        }
+//                    });
+//                }
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: Don't forget to remove

File: pos_debt_notebook/static/src/js/pos.js [hunk 3]
```diff
@@ -782,8 +781,8 @@ odoo.define('pos_debt_notebook.pos', function (require) {
         render_change: function() {
             // deduct the number of discount credits, because they were spent on discounts, but still presence like unspent
             var order = this.pos.get_order();
-            var change = order.get_change() - Math.max(order.get_summary_for_discount_credits(), 0);
-            this.$('.change-value').html(this.format_currency(change));
+//            var change = order.get_change() - Math.max(order.get_summary_for_discount_credits(), 0);
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: don't forget to remove

File: pos_debt_notebook/static/src/xml/pos.xml [hunk 4]
```diff
@@ -255,7 +255,8 @@
             <t t-set="credits_via_discounts_amount" t-value="Math.max(order.get_summary_for_discount_credits(), 0)"/>
             <table class='receipt-change'>
                 <tr><td>Change:</td><td class="pos-right-align">
-                    <t t-esc="widget.format_currency(order.get_change() - credits_via_discounts_amount)"/>
+                    <!--<t t-esc="widget.format_currency(order.get_change() - credits_via_discounts_amount)"/>-->
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/xml/pos.xml]: don't forget to remove

File: pos_debt_notebook/models.py [hunk 5]
```diff
@@ -482,14 +485,46 @@ def _process_order(self, pos_order):
...
+            price = line['qty'] * (unit_price + tax_summary)
+            old_price = price * (1 - line['discount'] / 100)
+            if not old_price:
+                continue
+            line['discount'] = round(max(min(line['discount'] + (amount / price) * 100, 100), 0), 2)
+            new_price = price * (1 - line['discount'] / 100)
+            line['price_unit'] += tax_summary
+            amount -= old_price - new_price
+            # taxes are already token into account in the product price
+            line['tax_ids'] = False
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: Can we use some built-in methods for this?
E.g. https://github.com/odoo/odoo/blob/11.0/addons/sale/models/sale.py#L1154-L1174
or (better)

https://github.com/odoo/odoo/blob/11.0/addons/point_of_sale/models/pos_order.py#L25-L31
https://github.com/odoo/odoo/blob/11.0/addons/point_of_sale/models/pos_order.py#L504-L513
https://github.com/odoo/odoo/blob/11.0/addons/point_of_sale/models/pos_order.py#L967-L976
https://github.com/odoo/odoo/blob/11.0/addons/point_of_sale/models/pos_order.py#L993-L1003

**Review by yelizariev — CHANGES_REQUESTED** [review]
set_discounts looks frighteningly

File: pos_debt_notebook/models.py [hunk 6]
```diff
@@ -482,14 +486,36 @@ def _process_order(self, pos_order):
...
+                break
+            price = line.price_subtotal_incl
+            old_price = price * (1 - line.discount / 100)
+            if not old_price:
+                continue
+            line.write({
+                'discount': round(max(min(line.discount + (amount / price) * 100, 100), 0), 2),
+            })
+            new_price = price * (1 - line.discount / 100)
+            amount -= old_price - new_price
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: Can this be done via built-in methods too?
https://github.com/it-projects-llc/pos-addons/pull/667#discussion_r183782447

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_debt_notebook/models.py (hunk 6 above)
Comment (KolushovAlexandr) [reply pos_debt_notebook/models.py]: @yelizariev I haven't seen any methods in the built-in code for recalculating absolute discount into a relative one

Comment (yelizariev) [reply pos_debt_notebook/models.py]: ok, got it

File: pos_debt_notebook/models.py [hunk 7]
```diff
@@ -482,14 +485,42 @@ def _process_order(self, pos_order):
...
+        for line in self.lines:
+            if abs(amount) < 0.00001:
+                break
+            price = line.price_subtotal_incl
+            if not price:
+                continue
+            disc = line.discount
+            full_price = disc and (price / (100 - disc)) * 100 or price
+            line.write({
+                'discount': round(max(min(line.discount + (amount / full_price) * 100, 100), 0), 2),
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: Use round methods from odoo framework

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_debt_notebook/models.py [hunk 8]
```diff
@@ -482,14 +485,44 @@ def _process_order(self, pos_order):
...
+                break
+            price = line.price_subtotal_incl
+            if not price:
+                continue
+            disc = line.discount
+            full_price = disc and (price / (100 - disc)) * 100 or price
+            currency = self.pricelist_id.currency_id
+
+            line.write({
+                'discount': currency.round(max(min(line.discount + (amount / full_price) * 100, 100), 0)),
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: default

File: pos_debt_notebook/models.py [hunk 9]
```diff
@@ -482,14 +485,44 @@ def _process_order(self, pos_order):
...
+    @api.model
+    def _order_fields(self, ui_order):
+        res = super(PosOrder, self)._order_fields(ui_order)
+        res['amount_via_discount'] = ui_order['amount_via_discount']
+        return res
+
+    def action_pos_order_paid(self):
+        self.set_discounts(self.amount_via_discount)
+        rounding_diff = self.amount_total - self.amount_paid
+        if abs(rounding_diff) > 0.00001:
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: use ``float_is_zero``

File: pos_debt_notebook/models.py [hunk 10]
```diff
@@ -482,14 +485,44 @@ def _process_order(self, pos_order):
...
+            self.set_discounts(rounding_diff)
+        return super(PosOrder, self).action_pos_order_paid()
+
+    def set_discounts(self, amount):
+        for line in self.lines:
+            if abs(amount) < 0.00001:
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: use ``float_is_zero``

File: pos_debt_notebook/models.py [hunk 11]
```diff
@@ -482,14 +485,44 @@ def _process_order(self, pos_order):
...
+                break
+            price = line.price_subtotal_incl
+            if not price:
+                continue
+            disc = line.discount
+            full_price = disc and (price / (100 - disc)) * 100 or price
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: Put it directly to discount formula

File: pos_debt_notebook/models.py [hunk 12]
```diff
@@ -482,14 +485,44 @@ def _process_order(self, pos_order):
...
+            currency = self.pricelist_id.currency_id
+
+            line.write({
+                'discount': currency.round(max(min(line.discount + (amount / full_price) * 100, 100), 0)),
+            })
+            new_price = price * (1 - line.discount / 100)
+            amount -= price - new_price
```
Comment (yelizariev) [inline pos_debt_notebook/models.py]: Try to use line.price_subtotal_incl:

`amount -= price - line.price_subtotal_incl`

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_debt_notebook/static/src/js/pos.js [hunk 13]
```diff
@@ -380,6 +379,20 @@ odoo.define('pos_debt_notebook.pos', function (require) {
                 });
                 return;
             }
+            var discount_pls = this.discount_credits_for_taxed_products()
+            if (discount_pls.length){
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: Make ``discount_pls``  boolean and don't use ``discount_pls.length``

**Review by yelizariev — CHANGES_REQUESTED** [review]
It's wrong check in general. It should be:

* if discount is not 100% (DONE)
* AND if product doesn't have taxes that are not included in the price (TODO)

File: pos_debt_notebook/static/src/js/pos.js [hunk 14]
```diff
@@ -506,6 +502,20 @@ odoo.define('pos_debt_notebook.pos', function (require) {
...
+            if(discount_pl.length) {
+                var check = discount_pl.length !== order.get_paymentlines().length && _.find(orderlines, function(ol){
+                    return ol.product.taxes_id && ol.product.taxes_id.length;
+                });
+                if(check) {
+                    return discount_pl;
+                }
+            }
+            return false;
+        },
```
Comment (yelizariev) [inline pos_debt_notebook/static/src/js/pos.js]: Split in two checks and rename method:
```

check_....: function(){
    ...
   if (has_only_payments_via_discount)
       return OK
   if (!has_products_with_not_included_taxes)
       return OK
   return ERROR
}
```

**Review by yelizariev — CHANGES_REQUESTED** [review]

**Review by yelizariev — APPROVED** [review]

**Pull request merged**